### PR TITLE
For #7725 - Log FxaExceptions to our crash reporters if sending to device fails

### PR DESCRIPTION
**Pull request by Mugurell**
This brings parity with iOS which already is logging such exceptions (see https://github.com/mozilla-mobile/fenix/issues/10662)

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/concept/sync/src/main/java/mozilla/components/concept/sync/Devices.kt
Comment (mhammond) [inline components/concept/sync/src/main/java/mozilla/components/concept/sync/Devices.kt:72]: It might be worth clarifying that it will (should?) only throw `FxaException` exceptions (here and below)

**Review by mhammond — APPROVED** [review]
This looks good to me, thanks! But there will need to be a matching PR for Fenix before it lands, otherwise I fear these now unhandled exceptions will cause Fenix to crash. I don't know enough about Kotlin to know whether that Fenix caller will be able to examine the exceptions correctly (ie, whether it will know the exact types so it can differentiate them correctly for Sentry) so it's probably important we know for sure this approach will work OK. So I'll approve this, but please don't land it before someone (eg, @liuche, or even me in a pinch ;) has approved the matching Fenix PR.

**Comment by Mugurell** [comment]
Thanks @mhammond 
The Fenix ticket already has a PR waiting - https://github.com/mozilla-mobile/fenix/pull/14810

File: components/concept/sync/src/main/java/mozilla/components/concept/sync/Devices.kt
Comment (Mugurell) [reply components/concept/sync/src/main/java/mozilla/components/concept/sync/Devices.kt]: Thanks!
Changed `Exception` occurrences to `FxaException`.

**Comment by mhammond** [comment]
I think the assignee should be @Mugurell ?

**Comment by Mugurell** [comment]
@mhammond @grigoryk I rebased the patch after the recent "Account manager state machine refactoring".
Would be great if you could take another look to confirm everything seems ok.
I tested and on the Fenix side everything is good.

File: components/feature/accounts-push/src/main/java/mozilla/components/feature/accounts/push/SendTabUseCases.kt
Comment (mhammond) [inline components/feature/accounts-push/src/main/java/mozilla/components/feature/accounts/push/SendTabUseCases.kt:132]: is this 'return' redundant?

**Review by mhammond — COMMENTED** [review]
This looks right to me, but please get Grisha's review!

File: components/feature/accounts-push/src/main/java/mozilla/components/feature/accounts/push/SendTabUseCases.kt
Comment (Mugurell) [reply components/feature/accounts-push/src/main/java/mozilla/components/feature/accounts/push/SendTabUseCases.kt:132]: It is, thanks!

**Comment by liuche** [comment]
@st3fan could you get this pushed through? It's been open for a while.

**Comment by st3fan** [comment]
I would like to ask @grigoryk for an additional review of this patch

**Comment by st3fan** [comment]
@Mugurell can you bring this patch up to date?

**Comment by liuche** [comment]
Hey @grigoryk @jonalmeida for the send tab issues that y'all are running into, can we get this PR reviewed and landed? It might have helpful logs as well.

The only conflicts I see are in the changelog so it's ready for review.

**Comment by grigoryk** [comment]
@liuche taking a look at both PRs now. I don't think they'll help with debugging our current problems though (since this is dealing with the Send part, which works well atm).

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/Utils.kt
Comment (grigoryk) [inline components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/Utils.kt:81]: If your plan is to just re-throw whatever it is `block` throws, you don't need this wrapper at all. Just execute `block` directly!

**Review by grigoryk — CHANGES_REQUESTED** [review]
While this approach works, I don't think it's the one we should take. Specifically, forcing consumers to deal with FxaExceptions is a potential foot gun (they must handle auth errors correctly, re-throw panics, swallow network errors, etc), leaks a pretty low-level detail all the way to the top of our stack, and ultimately complicates our implementation of these APIs.

The problem you're trying to solve is that a `boolean` result for `sendCommandToDevice` doesn't provide enough information about what actually happened, making it difficult to log an exact failure to Sentry. You can solve that by introducing a richer result, e.g.:

```
enum class SendCommandResult {
  Success,
  NetworkFailure,
  AuthFailure,
  OtherFailure
}
```

Or consider a `sealed class` if you'd like to associate more data with these exceptions, e.g.:
```
sealed class SendCommandResult {
   object Success : SendCommandResult()
   sealed class Failure : SendCommandResult() {
      data class Network(val code: Int) : Failure()
      object AuthFailure : Failure()
      data class OtherFailure(val details: String?) : Failure()
   }
}
```

You will then convert an FxaException into one of these, _while also_ letting `handleFxaException` do its thing - e.g. inform the state machine of auth failures, re-throw panics (these _will be_ reported to Sentry as uncaught crashes).

Optionally within the `SendToAllUseCase` class, if you choose, you may aggregate these results into something like the following:

```
enum class SendCommandToAllResult {
  Success,
  PartialFailure,
  Failure
}
```

But to be honest, I think that's unnecessary. We can infer health of this feature by looking at individual send results.

In order to report these to Sentry, you should create a new Exception class wrapping your `Failure`:

`class SendCommandException(val failure: SendCommandResult.Failure) : Exception(failure.toString())`

and then submit that to Sentry.

All of these changes can be contained entirely within `FxaDeviceConstellation` - it already has the `crashReporter` instance.

This way you won't need any changes in Fenix, and you don't even have to change any public API - e.g. you can just keep the `Boolean` interface externally, and not expose the internal rich `SendTabResult` at all, just use that for detailed error reporting.

... or you can expose it to consumers, with a hope that they can then build a nicer UI (e.g. Fenix will be able to call out a network failure in a Snackbar notification! that seems quite nice). But, consider doing that in a follow-up, e.g. this isn't central to your goal of better error reporting via Sentry.

**Comment by Mugurell** [comment]
Thank you @grigoryk !

I now have `handleFxaExceptions` returning a `FxaException?` which when in `sendCommandToDevice` I just wrap in a `SendCommandException` and submit it to the crash reporter.

The `FxaException?` coming from `handleFxaExceptions` already has the appropriate subtypes for better identifying the exception cause so I didn't create another `SendCommandResult`. I can also wrap these FxaExceptions in new `SendCommandResult` or just map their stacktrace to new appropriate subtypes if you think this is needed.

**Comment by Mugurell** [comment]
> @liuche taking a look at both PRs now. I don't think they'll help with debugging our current problems though (since this is dealing with the Send part, which works well atm).

The intent for this, based on https://github.com/mozilla-mobile/fenix/issues/10662 was just to have parity with iOS which would log such exceptions, if they were to happen.

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/Utils.kt
Comment (grigoryk) [inline components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/Utils.kt:77]: You don't need to change this method. Instead, in `sendCommandToDevice` change how `handleFxaExceptions` is used:

```
        val result = handleFxaExceptions(logger, "sending device command", { error -> error }) {
            when (outgoingCommand) {
                is DeviceCommandOutgoing.SendTab -> {
                    account.sendSingleTab(targetDeviceId, outgoingCommand.title, outgoingCommand.url)
                    SyncTelemetry.processFxaTelemetry(account.gatherTelemetry(), crashReporter)
                }
                else -> logger.debug("Skipped sending unsupported command type: $outgoingCommand")
            }
            null
        }
```

And that should be pretty much it!

**Review by grigoryk — COMMENTED** [review]
Nice! You can simplify this even further, see my comment.

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/Utils.kt
Comment (Mugurell) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/Utils.kt:77]: Thanks!

**Review by grigoryk — APPROVED** [review]
Looks good, thanks!

**Pull request merged**
